src/parser.ts

```diff
@@ -1,4 +1,152 @@
-import { ValidationError } from "./common";
+/**
+ * The parser is responsible for taking a raw JS object
+ * of the XTP Schema and parsing and typing it without transforming
+ * any of the values. It's just mean to give a typed representation
+ * of the raw format.
+ */
+import { ValidationError } from "./common"
+
+export interface ParseResult {
+  doc?: VUnknownSchema;
+  errors?: ValidationError[];
+}
+
+/**
+ * Parses and validates an untyped object into a V*Schema
+ */
+export function parseAny(doc: any): ParseResult {
+  switch (doc.version) {
+    case 'v0':
+      return { doc: doc as V0Schema }
+    case 'v1-draft':
+      const v1Doc = doc as V1Schema
+      const validator = new V1Validator(v1Doc)
+      const errors = validator.validate()
+      return { doc: v1Doc, errors }
+    default:
+      return {
+        errors: [
+          new ValidationError(`version property not valid: ${doc.version}`, "#/version")
+        ]
+      }
+  }
+}
+
+export function isV0Schema(schema?: VUnknownSchema): schema is V0Schema {
+  return schema?.version === 'v0';
+}
+
+export function isV1Schema(schema?: VUnknownSchema): schema is V1Schema {
+  return schema?.version === 'v1-draft';
+}
+
+/**
+ * Validates a V1 document.
+ */
+class V1Validator {
+  errors: ValidationError[]
+  location: string[]
+  doc: any
+
+  constructor(doc: V1Schema) {
+    this.doc = doc as any
+    this.errors = []
+    this.location = ['#']
```

Comment:
instead of hand building the location we can just track it as we walk the nodes.

Reply:
beautiful!